WordPress/gutenberg
# E2E Utils: add support for web-vitals.js

## What?


This adds support for getting web vitals measurements via the [web-vitals](https://github.com/GoogleChrome/web-vitals) library to `@wordpress/e2e-test-utils-playwright`.

## Why?


The existing `Metrics` fixture already provides a handful of useful helper methods such as `getLargestContentfulPaint` or `getCumulativeLayoutShift` to get some web vitals values. These are based on recommended techniques and existing APIs. 

However, this doesn't scale. There are always nuances to how things are measured, and the metrics evolve all the time. Manually maintaining a way to measure INP for example would be too cumbersome over time.

With the web-vitals library we get all of it out of the box with very little work. We could even consider removing `getLargestContentfulPaint`, `getCumulativeLayoutShift`, and `getTimeToFirstByte` in favor of `getWebVitals()` so there is only one API surface for the same metrics.

This partially unblocks tickets such as core-59119 and core-59120.

## How?


Uses the library's UMD build and exposes a special `__reportVitals__` function to the page to directly add any received metrics to the class property.

By automatically handling things like loading the library and doing page navigation, usage is very simple, yet it also allows for more manual control.

**TypeScript**

- `LayoutShift` and `LayoutShiftAttribution` types no longer need to be provided manually. They are provided by `web-vitals`
- Had difficulties typing the `webVitals` global as set up by the UMD build. Hence the `@ts-ignore` comments.

## Testing Instructions

Follow the example provided in the docblock. For example adding this to `test/performance/specs/front-end-classic-theme.spec.js` successfully reports web vitals metrics for the home page:

```js
// Go to the base URL.
await page.goto( '/' );

console.log( await metrics.getWebVitals() );
```

Note that you won't see any INP values in there because you need to actually add some user interaction for that.
